File: app/src/main/java/me/tysheng/xishi/adapter/BaseGalleryAdapter.java
package me.tysheng.xishi.adapter;

import android.app.Activity;
import android.support.v4.view.PagerAdapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

import me.tysheng.xishi.R;
import uk.co.senab.photoview.PhotoViewAttacher;

/**
 * Created by shengtianyang on 16/4/30.
 */
public abstract class BaseGalleryAdapter<T> extends PagerAdapter {
    protected List<T> mImages;
    protected Activity mActivity;

    public BaseGalleryAdapter(List<T> images, Activity activity) {
        mImages = images;
        mActivity = activity;
    }

    public void setData(List<T> images) {
        mImages = images;
        notifyDataSetChanged();
    }

    @Override
    public Object instantiateItem(ViewGroup container, final int position) {
        ViewGroup view = (ViewGroup) LayoutInflater.from(mActivity).inflate(setLayoutId(), container, false);
        final PhotoViewAttacher mAttacher;
        final ImageView imageView = (ImageView) view.findViewById(setImageViewId());
        final ProgressBar progressBar = (ProgressBar) view.findViewById(setProgressBarId());
        initOtherView(view, position);
        mAttacher = new PhotoViewAttacher(imageView);
        mAttacher.setScaleType(ImageView.ScaleType.FIT_CENTER);
        mAttacher.setMinimumScale(1);
        initAttacher(mAttacher, position);
        Picasso.with(mActivity)
                .load(setItemUrl(position))
                .error(R.layout.item_loading)
                .into(imageView, new Callback() {
                    @Override
                    public void onSuccess() {
                        progressBar.setVisibility(View.GONE);
                        mAttacher.update();
                    }

                    @Override
                    public void onError() {

                    }
                });
        container.addView(view);
        return view;
    }

    protected abstract void initOtherView(View view, int position);

    protected abstract void initAttacher(PhotoViewAttacher attacher, int position);

    protected abstract String setItemUrl(int position);

    protected abstract int setLayoutId();

    protected abstract int setImageViewId();

    protected int setProgressBarId() {
        return R.id.progressBar;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        View view = (View) object;
        container.removeView(view);
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return setPageTitle(position);
    }

    protected abstract CharSequence setPageTitle(int position);

    @Override
    public int getCount() {
        return mImages.size();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

}

File: app/src/main/java/me/tysheng/xishi/adapter/BaseGalleryAdapter.kt
package me.tysheng.xishi.adapter

import android.content.Context
import android.support.v4.view.PagerAdapter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import com.github.chrisbanes.photoview.PhotoView
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import me.tysheng.xishi.R

/**
 * Created by shengtianyang on 16/4/30.
 */
abstract class BaseGalleryAdapter<T>(protected var images: List<T>
                                     , protected var context: Context) : PagerAdapter() {

    var data: List<T>
        get() = images
        set(images) {
            this.images = images
            notifyDataSetChanged()
        }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val view = LayoutInflater.from(context).inflate(setLayoutId(), container, false) as ViewGroup
        val imageView = view.findViewById<View>(R.id.imageView) as PhotoView
        val progressBar = view.findViewById<View>(setProgressBarId()) as ProgressBar
        initOtherView(view, position)
        imageView.minimumScale = 1f
        Picasso.get()
                .load(setItemUrl(position))
                .into(imageView, object : Callback {
                    override fun onSuccess() {
                        progressBar.visibility = View.GONE
                    }

                    override fun onError(e: Exception) {

                    }
                })
        container.addView(view)
        return view
    }

    protected open fun initOtherView(view: View, position: Int){

    }

    protected open fun initPhotoView(photoView: PhotoView, position: Int){

    }

    protected abstract fun setItemUrl(position: Int): String

    protected abstract fun setLayoutId(): Int

    protected open fun setProgressBarId(): Int {
        return R.id.progressBar
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        val view = `object` as View
        container.removeView(view)
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return setPageTitle(position)
    }

    protected abstract fun setPageTitle(position: Int): CharSequence

    override fun getCount(): Int {
        return images.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

}
